flowd-reader.8.in
.\" $Id: flowd-reader.8.in,v 1.6 2005/09/02 10:44:26 djm Exp $
.\"
.\" Copyright (c) 2004 Damien Miller <djm@mindrot.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd July 30, 2004
.Dt FLOWD-READER 8
.Os
.Sh NAME
.Nm flowd-reader
.Nd Read, filter and concatenate binary flowd logfiles
.Sh SYNOPSIS
.Nm flowd-reader
.Op Fl LUvqd
.Op Fl H Ar num_flows
.Op Fl f Ar filter_file
.Op Fl o Ar output_file
.Ar flow_log
.Op Ar flow_log
.Op Ar ...
.Sh DESCRIPTION
.Nm
is a program to read, filter and/or concatenate
.Xr flowd 8
binary log files.
.Pp
If only
.Ar flow_log
arguments are specified,
.Nm
will read each of the log files and print their contents to standard output.
.Pp
Filtering may be performed by specifying a 
.Ar filter_file
using the
.Fl f
option.
In this mode,
.Nm
will filter the flows it reads from the 
.Ar flow_log
file(s) before printing them.
.Pp
Finally, the
.Fl o
option may be used to specify a
.Ar output_path
to which all the flows that have been read and passed any filters are written.
.Pp
The command-line options are as follows:
.Bl -tag -width Ds
.It Fl H Ar num_flows
.Xr head 1
mode.
Read only the first
.Ar num_flows
of the file.
.It Fl L
Allows
.Nm
to read legacy version 2 flow logs (generated by
.Xr flowd 8
versions prior to v9.0).
This may be used to convert old flow logs to the newer form.
.It Fl U
Causes
.Nm
to report all timestamps in UTC rather than the local timezone.
.It Fl d
Display debugging information, including the number of filter matches if one 
has been specified.
.It Fl f Ar filter_file
Specify a 
.Ar filter_file .
containing filtering and field selection rules to apply to the flows that are
loaded.
These rules may contain 
.Ar store
directives and filter rules as described in the 
.Xr flowd.conf 5
manual (under the
.Cm Storage field selection
and 
.Cm Filter
sections respectively.)
Any other directives are ignored.
If no
.Ar store
directives are specified in the 
.Ar filter_file
then the default is to preserve all the fields in the input flow logs.
.It Fl q
Operate quietly. If this argment is specified,
.Nm
will not display the flows it is processing.
This may be useful in conjunction with the
.Fl f
option.
.It Fl o Ar output_file
Specify an
.Ar output_file
to which all flows that have been read and have passed any filters are written.
The flows are written in the
.Nm flowd
binary log format.
This option is useful when filtering or concatenating flow log files.
.It Fl v
Reports all information in the flow log, rather than the default brief subset.
.It Fl h
Displays commandline usage information.
.El
.Sh AUTHORS
Damien Miller <djm@mindrot.org>
.Sh SEE ALSO
.Xr flowd 8 ,
.Xr flowd.conf 5


ChangeLog
20080725
 - (djm) byte swap interface indices before testing them in filters
 - (djm) Fixed swapped last/first_switched times in netflow v.9 code
   patch from weinhold AT berbee.com
 - (djm) Released flowd-0.9.1

20080422
 - (djm) Support flow-tools CSV output format in flowd-reader. Patch from
   weinhold AT berbee.com
 - (djm) Use proper API for logging debug information; Spotted by
   JSaxe AT briworks.com

20071024
 - (djm) Support explicit specification of "listen on" and "logsock"
   socket buffer sizes. Patch from kempfj2 AT cs.rpi.edu

20071010
 - (djm) Make local socket-only logging work, and fix spurious warnings
   from socket logging. Patch from kempfj2 AT cs.rpi.edu
 - (djm) Unbreak flowinsert.pl

20070723
 - (djm) Fix NetFlow v.9 flowset template parsing on LP!32 platforms.
   Report and fix from zhangjinxue AT cernet.edu.cn

20070510
 - (djm) Add support for filtering input/output interface index. Requested
   by Ralf Kleineisel, debugging assistance from Matthew Smart

20061020
 - (djm) Rename CRC32 functions to avoid collisions with zlib;
   patch from Sergey Vasilenko <stalker@telemost.ru>

20051227
 - (djm) Add RPM spec and init files for SuSE Linux from alshu AT tut.by
 - (djm) Released flowd-0.9

20051220
 - (djm) Fix byte swapping of src/dst AS and interface indices, spotted 
   and fix tested by Gijs Molenaar

20051208
 - (djm) Implement absolute time (before & after) filters, requested by
   alshu AT tut.by

20051207
 - (djm) Demote input buffer full warning to a debug and reduce the length
   of the input queue (now that we have an output queue)
 - (djm) Another hex that got away

20051206
 - (djm) Rename some filter members to make absolute time filtering easier

20051204
 - (djm) Add an output queue, so we don't do tiny little filesystem writes for
   each flow we receive
 - (djm) Sync tree.h and queue.h with OpenBSD

20051117
 - (djm) Prefix all hex integers in logs with "0x"
 - (djm) More source_id debugging everywhere

20051111
 - (djm) Support devices that send multiple templates in a single template
   packet section. Thanks to Gijs Molenaar for packet dumps that demonstrated
   this
 - (djm) Turn off Netflow v.9 debugging

20051013
 - (djm) Fix bug that broke filtering on address family in flows, spotted by
   Gijs Molenaar
 - (djm) Rename common.h -> flowd-common.h, always install it and
   flowd-config.h. Should help platforms without intXX_t etc.

20051001
 - (djm) Add FlowLog_from_file method to Python module
 - (djm) Increase UDP socket receive buffer size, shrink socket send buffer size
   to cope better with bursts of flows
 - (djm) Allow reading from standard input in tools/stats.py
 - (djm) Better error message on corrupt flows
 - (djm) Change flowd-reader to *overwrite* existing log files, not *append* to
   them

20050928
 - (djm) Add a basic input queue to flowd, to improve its behaviour when it
   receives sudden bursts of packets.

20050918
 - (djm) Fix error reporting on logsock send failures
 - (djm) Fix another typo in tools/stats.py, spotted by alshu@tut.by
 - (djm) Sync usage for flowd-reader
 - (djm) -Wall cleanup flowd_python.c
 - (djm) Tidy stats.py, and remove hard dependency on curses module

20050918
 - (djm) Fix typo in tools/stats.py, spotted by alshu@tut.by
 - (djm) Mention softflowd and pfflowd in README, suggested by alshu@tut.by

20050913
 - (djm) Mention creation of _flowd user in INSTALL, spotted by alshu@tut.by

20050826
 - (djm) Add interval_time() and iso_time() functions to Python module
 - (djm) Port and tidy stats script
 - (djm) Add a "head" mode to flowd-reader

20050826
 - (djm) Add Flow.has_field() to Python API

20050825
 - (djm) Make it compile on Solaris 9 (Perl module is still busted because 
   of compiler misguessing stupidity)

20050824
 - (djm) Add support for relaying serialised flows to a local Unix domain
   datagram socket in realtime. This option ("logsock" in flowd.conf) is
   considered experimental. A sample Python client exists in 
   tools/sockclient.py that receives and prints the flows.

20050822
 - (djm) Improve the Python API some more:
   - Prefer PyLong as type for various fields
   - Maintain all addresses as PyObjects rather than char*, so it is 
     possible to write to them
   - Initialise missing addresses, octet and packet counters to Py_None
   - Fix lots of bugs
 - (djm) Macroise much of the NetFlow v.9 parsing code and fix length
   mismatches in the process
 - (djm) Check for, and refuse to append to legacy logfiles
 - (djm) Don't lookup username when running in INSECURE mode
 
20050822
 - (djm) Figure out Python structmember.h types for struct store_flow_complete
   at autoconf time and stuff them into their own header
 - (djm) Add (src|dst|agent|gateway)_addr_af to Python API

20050821
 - (djm) Major rewrite of storage code:
   - Introduce a new storage format (version 3), which should be much faster
     to read from disk and easier and more graceful to extend
   - Add FILE* oriented API
   - Add some new fields: receive time microseconds, netflow v.9 source_id
   - Extend the widths of several fields: if_ndx_(in|out), (src|dst)_as, 
     engine_(type|id)
   - Add functions for reading and writing legacy logs
 - (djm) Major rewrite of Python API
   - Implements all the functions in all hybrid API using 100% C code
   - Use Python structmember API and store.c for major speed improvement
   - Adds an iterator object to read all flows from a flow log
 - (djm) Adjust Perl API to cope with new store format
 - (djm) Add ability for flowd-reader to read and convert legacy log files
 - (djm) Update copyright years

20050707
 - (djm) Fix pidfile path in manpage

20050619
 - (djm) README typo, spotted by cruel AT texnika.com.ua

200500601
 - (djm) Add a simple Python summarisation and charting script (tools/stats.py)

20050514
 - (djm) Fix spurious error message
 - (djm) Improve day filters, allowing lists and ranges of days
 - (djm) Release flowd-0.8.5

20050428

flowd-reader.c
/*
 * Copyright (c) 2004,2005 Damien Miller <djm@mindrot.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#define PROGNAME	"flowd-reader"

#include "flowd-common.h"

#include <sys/types.h>

#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <poll.h>

#include "flowd.h"
#include "store.h"
#include "store-v2.h"
#include "atomicio.h"

RCSID("$Id: flowd-reader.c,v 1.23 2008/04/23 01:54:26 djm Exp $");

static void
usage(void)
{
	fprintf(stderr, "Usage: %s [options] flow-log [flow-log ...]\n",
	    PROGNAME);
	fprintf(stderr, "This is %s version %s. Valid commandline options:\n",
	    PROGNAME, PROGVER);
	fprintf(stderr, "  -H num   Read and/or write only the first 'num' flows\n");
	fprintf(stderr, "  -L       Read/convert legacy flow logs\n");
	fprintf(stderr, "  -q       Don't print flows to stdout (use with -o)\n");
	fprintf(stderr, "  -d       Print debugging information\n");
	fprintf(stderr, "  -f path  Filter flows using rule file\n");
	fprintf(stderr, "  -o path  Write binary log to path (use with -f)\n");
	fprintf(stderr, "  -v       Display all available flow information\n");
	fprintf(stderr, "  -c       Return CSV output compatible with flow-import\n");
	fprintf(stderr, "  -U       Report times in UTC rather than local time\n");
	fprintf(stderr, "  -h       Display this help\n");
}

static int
open_start_log(const char *path, int debug)
{
	int fd;

	if (path == NULL) {
		/* Logfile on stdout */
		fd = STDOUT_FILENO;
	} else if ((fd = open(path, O_RDWR|O_CREAT, 0600)) == -1)
		logerr("open(%s)", path);

	if (debug)
		fprintf(stderr, "Writing new logfile header\n");

	return (fd);
}


int
main(int argc, char **argv)
{
	int ch, i, fd, utc, r, verbose, debug, csv;
	extern char *optarg;
	extern int optind;
	struct store_flow_complete flow;
	struct store_v2_flow_complete flow_v2;
	char buf[2048], ebuf[512];
	const char *ffile, *ofile;
	FILE *ffilef;
	int ofd, read_legacy, head, nflows;
	u_int32_t disp_mask;
	struct flowd_config filter_config;
	struct store_v2_header hdr_v2;

	utc = verbose = debug = read_legacy = csv = 0;
	ofile = ffile = NULL;
	ofd = -1;
	ffilef = NULL;
	head = 0;

	bzero(&filter_config, sizeof(filter_config));

	while ((ch = getopt(argc, argv, "H:LUdf:ho:qvc")) != -1) {
		switch (ch) {
		case 'h':
			usage();
			return (0);
		case 'H':
			if ((head = atoi(optarg)) <= 0) {
				fprintf(stderr, "Invalid -H value.\n");
				usage();
				exit(1);
			}
			break;
		case 'L':
			read_legacy = 1;
			break;
		case 'U':
			utc = 1;
			break;
		case 'd':
			debug = 1;
			filter_config.opts |= FLOWD_OPT_VERBOSE;
			break;
		case 'f':
			ffile = optarg;
			break;
		case 'o':
			ofile = optarg;
			break;
		case 'q':
			verbose = -1;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'c':
			csv = 1;
			break;
		default:
			usage();
			exit(1);
		}
	}
	loginit(PROGNAME, 1, debug);

	if (argc - optind < 1) {
		fprintf(stderr, "No logfile specified\n");
		usage();
		exit(1);
	}

	if (ffile != NULL) {
		if ((ffilef = fopen(ffile, "r")) == NULL)
			logerr("fopen(%s)", ffile);
		if (parse_config(ffile, ffilef, &filter_config, 1) != 0)
			exit(1);
		fclose(ffilef);
	}

	if (ofile != NULL) {
		if (strcmp(ofile, "-") == 0) {
			if (!debug)
				verbose = -1;
			ofile = NULL;
			if (isatty(STDOUT_FILENO))
				logerrx("Refusing to write binary flow data to "
				    "standard output.");
		}
		ofd = open_start_log(ofile, debug);
	}

	if (filter_config.store_mask == 0)
		filter_config.store_mask = STORE_FIELD_ALL;

	disp_mask = (verbose > 0) ? STORE_DISPLAY_ALL: STORE_DISPLAY_BRIEF;
	disp_mask &= filter_config.store_mask;

	for (i = optind; i < argc; i++) {
		if (strcmp(argv[i], "-") == 0)
			fd = STDIN_FILENO;
		else if ((fd = open(argv[i], O_RDONLY)) == -1)
			logerr("open(%s)", argv[i]);

		if (read_legacy && store_v2_get_header(fd, &hdr_v2, ebuf,
		    sizeof(ebuf)) != STORE_ERR_OK)
			logerrx("%s", ebuf);

		if (verbose >= 1) {
			printf("LOGFILE %s", argv[i]);
			if (read_legacy)
				printf(" started at %s",
				    iso_time(ntohl(hdr_v2.start_time), utc));
			printf("\n");
			fflush(stdout);
		}

		if (csv == 1) {
			csv++;
			printf("#:unix_secs,unix_nsecs,sysuptime,exaddr,"
			    "dpkts,doctets,first,last,engine_type,engine_id,"
			    "srcaddr,dstaddr,nexthop,input,output,srcport,"
			    "dstport,prot,tos,tcp_flags,src_mask,dst_mask,"
			    "src_as,dst_as\n");
		}

		for (nflows = 0; head == 0 || nflows < head; nflows++) {
			bzero(&flow, sizeof(flow));

			if (read_legacy)
				r = store_v2_get_flow(fd, &flow_v2, ebuf,
				    sizeof(ebuf));
			else
				r = store_get_flow(fd, &flow, ebuf,
				    sizeof(ebuf));
				
			if (r == STORE_ERR_EOF)
				break;
			else if (r != STORE_ERR_OK)
			    	logerrx("%s", ebuf);

			if (read_legacy &&
			    store_v2_flow_convert(&flow_v2, &flow) == -1)
			    	logerrx("legacy flow conversion failed");

			if (ffile != NULL && filter_flow(&flow,
			    &filter_config.filter_list) == FF_ACTION_DISCARD)
				continue;
			if (csv) {
				store_format_flow_flowtools_csv(&flow, buf,
				    sizeof(buf), utc, disp_mask, 0);
				printf("%s\n", buf);
			}
			else if (verbose >= 0) {
				store_format_flow(&flow, buf, sizeof(buf), 
				    utc, disp_mask, 0);
				printf("%s\n", buf);
				fflush(stdout);
			}
			if (ofd != -1 && store_put_flow(ofd, &flow, 
			    filter_config.store_mask, ebuf, 
			    sizeof(ebuf)) == -1)
			    	logerrx("%s", ebuf);
		}
		if (fd != STDIN_FILENO)
			close(fd);
	}
	if (ofd != -1)
		close(ofd);

	if (ffile != NULL && debug)
		dump_config(&filter_config, "final", 1);

	return (0);
}
